include/wil/safearrays.h

```diff
@@ -0,0 +1,897 @@
+        template<> struct var_traits<float>                      { static constexpr auto vartype = VT_R4; };
+        //template<> struct var_traits<double>                     { static constexpr auto vartype = VT_R8; };
+        template<> struct var_traits<VARIANT_BOOL>               { static constexpr auto vartype = VT_BOOL; };
+        template<> struct var_traits<DATE>                       { static constexpr auto vartype = VT_DATE; };
+        template<> struct var_traits<CURRENCY>                   { static constexpr auto vartype = VT_CY; };
+        template<> struct var_traits<DECIMAL>                    { static constexpr auto vartype = VT_DECIMAL; };
+        template<> struct var_traits<BSTR>                       { static constexpr auto vartype = VT_BSTR; };
+        template<> struct var_traits<LPUNKNOWN>                  { static constexpr auto vartype = VT_UNKNOWN; };
+        template<> struct var_traits<LPDISPATCH>                 { static constexpr auto vartype = VT_DISPATCH; };
+        template<> struct var_traits<VARIANT>                    { static constexpr auto vartype = VT_VARIANT; };
+
+        inline void __stdcall SafeArrayDestroy(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            FAIL_FAST_IF_FAILED(::SafeArrayDestroy(psa));
+        }
+
+        inline void __stdcall SafeArrayLock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            FAIL_FAST_IF_FAILED(::SafeArrayLock(psa));
+        }
+
+        inline void __stdcall SafeArrayUnlock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            FAIL_FAST_IF_FAILED(::SafeArrayUnlock(psa));
+        }
+
+        template<typename T>
+        inline void __stdcall SafeArrayAccessData(SAFEARRAY* psa, T*& p) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            FAIL_FAST_IF_FAILED(::SafeArrayAccessData(psa, reinterpret_cast<void**>(&p)));
+        }
+
+        inline void __stdcall SafeArrayUnaccessData(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            __FAIL_FAST_ASSERT__(psa != nullptr);
+            FAIL_FAST_IF_FAILED(::SafeArrayUnaccessData(psa));
+        }
+
+        inline VARTYPE __stdcall SafeArrayGetVartype(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            VARTYPE vt = VT_NULL;   // Invalid for SAs, so use to mean SA was null
+            if ((psa != nullptr) && FAILED(::SafeArrayGetVartype(psa, &vt)))
+            {
+                vt = VT_EMPTY;  // Invalid for SAs, use to mean type couldn't be determined
+            }
+            return vt;
+        }
+
+        inline ULONG __stdcall SafeArrayGetLockCount(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            return (psa != nullptr) ? psa->cLocks : 0U;
+        }
+
+        inline HRESULT __stdcall SafeArrayCreate(VARTYPE vt, UINT cDims, SAFEARRAYBOUND* sab, SAFEARRAY*& psa) WI_NOEXCEPT

```

Comment:
Don't use hungarian

Reply:
I tried to use the same variable names as in the API documentation, which makes it easier to understand for people already familiar with the documentation, but I'll change it.